aiida_vasp/workchains/vasp.py

```diff
@@ -455,20 +455,41 @@ def _handle_unfinished_calculation(self, node):
                     exit_code=self.exit_codes.ERROR_MANUAL_INTERVENTION_NEEDED.format(message='No WAVECAR for continuation.'))  #pylint: disable=no-member
         return ProcessHandlerReport(do_break=True)
 
-    @process_handler(priority=800, exit_codes=[VaspCalculation.exit_codes.ERROR_ELECTRONIC_NOT_CONVERGED])
+    @process_handler(priority=800,
+                     exit_codes=[
+                         VaspCalculation.exit_codes.ERROR_ELECTRONIC_NOT_CONVERGED,
+                         VaspCalculation.exit_codes.ERROR_IONIC_NOT_CONVERGED,
+                         VaspCalculation.exit_codes.ERROR_DID_NOT_FINISH,
+                     ])
     def _handle_electronic_converge_problem(self, node):
         """Handle electronic convergence problem"""
         incar = node.inputs.parameters.get_dict()
         run_status = node.outputs.misc['run_status']
         nelm = run_status['nelm']
         algo = incar.get('algo', 'normal')
+
+        # In case of ionic convergence problem, we only act where electronic convergence
+        # issues are consistently observed
+        # For finished runs - all steps are counted, and for unfinished runs all except the last step
+        # are included
+        # No action is taken for calculation did not finish the first cycle
+        if node.exit_status in [
+                VaspCalculation.exit_codes.ERROR_IONIC_NOT_CONVERGED.status, VaspCalculation.exit_codes.ERROR_DID_NOT_FINISH
+        ]:
+            if not run_status['consistent_nelm_breach']:
+                if run_status['contains_nelm_breach']:
+                    self.report('The NELM limit has been breached in some but not all ionic steps - no action taken.')
```

Comment:
I think, given the framework we have now that this should still be a hard fail as we can not discriminate between an almost converged and a far from converged result. Latter can give really wrong results and we should not support that out of the box.

What we might want to consider to implement is something that checks all steps this and then we can have some tolerance. But this is also a bit of a slippery slope.

We need to assume that when the result is not converged it is far from converged. Better to err on the same side.

Reply:
Well, the result is still both converged electronically (e.g. at a local minimum of the Born-Oppenheimer potential energy surface) as long as the last electronic cycle was converged and the ionic relaxation is converged. It may not be a nearby local minimum for some complicated cases if one of the unconverged electronic cycles gives inaccurate forces are that are too large. 

However, there is actually no guarantee that the optimiser will get us to the local minimum whose basin of attraction includes the starting structure even without any issues like this, for example, the very first move may just move the structure outside the basin.

The problem is that, if VASP wants to be very strict about this then it should stop if any of the electronic steps was not converged immediate, but it chose not to do this. Now we have a structure that is both at a local minimum and converged energy, but just one or two of the electronic cycle was not converged and all the effort 💰  has been done. I think the common practice (to me and other people in my group) is to just accept it as it is fine in most cases, which is also consistent with how VASP operates itself.

If you think this is really a killer, I can make it a separate handler so one can choose to operate in a "strict" mode.

Reply:
The point here is that we do not really know. Yes, we can make assumptions etc. and that is fine on a case-by-case basis by the person running the code. But that person should really want to do this and it should not be recommended or the default approach from the code. In fact, what should be done is to stop the calculation at the ionic step where `NELM` is reached. This is certainly something we will implement (yes, such a thing should be in VASP itself, but we would have to work with how it is now) when we can monitor jobs and the convergence. Then we can also truncate the calculations with the electronic convergence starts to fluctuate such that we do not waste resources. This is anyway something we would want in not too distant of a future.

I indeed think enabling this out of the box is not something we should do. I am of course happy to support optional functionality, such that the user can enable "do not care about NELM out of bounds in intermediate ionic steps", but that should be left in a disabled state by default. Also, it should be very clear what the user is doing when they enable it.

How each group/person wants to run the calcs and what approximations they want to introduce have to be up to them to choose.

Reply:
OK, I will add a flag to allow _do not care about NELM out of bounds in intermediate ionic steps_ with default to be off, so the check is strict.

Reply:
Awesome. That is great.